arangod/VocBase/Methods/UpgradeTasks.cpp

```diff
@@ -155,38 +120,372 @@ bool UpgradeTasks::upgradeGeoIndexes(TRI_vocbase_t& vocbase,
         if (res.fail()) {
           LOG_TOPIC("5550a", ERR, Logger::STARTUP)
               << "Error upgrading geo indexes " << res.errorMessage();
-          return false;
+          throw res;
         }
       }
     }
   }
+}
 
-  return true;
+Result createSystemCollections(TRI_vocbase_t& vocbase,
+                               std::vector<std::shared_ptr<LogicalCollection>>& createdCollections) {
+  typedef std::function<void(std::shared_ptr<LogicalCollection> const&)> FuncCallback;
+  FuncCallback const noop = [](std::shared_ptr<LogicalCollection> const&) -> void {};
+
+  std::vector<CollectionCreationInfo> systemCollectionsToCreate;
+  // the order of systemCollections is important. If we're in _system db, the
+  // UsersCollection needs to be first, otherwise, the GraphsCollection must be first.
+  std::vector<std::string> systemCollections;
+
+  Result res;
+  std::shared_ptr<LogicalCollection> colToDistributeShardsLike;
+  if (vocbase.isSystem()) {
+    // we will use UsersCollection for distributeShardsLike
+    std::tie(res, colToDistributeShardsLike) =
+        methods::Collections::createSystem(vocbase, StaticStrings::UsersCollection, true);
```

Comment:
I think this is not correct.
Assume we have an older ArangoDB where users is existing.
In this case the command here will fail and the following collections will not be created.
Instead we need to check for existence of `_users` if it is there, we need to use it's distribute shards like (if set) or we distribute by it.

Reply:
Yes, this is important. This suggests that it would probably have been easier altogether to stick with everything distributeShardsLike `_graphs`. Yes, we would have to create the `_graphs` collection without adapting its user rights, but this seems overall a smaller "Extrawurst". Too late to go back?

Reply:
It does check: 

See: 
`  Result res = methods::Collections::lookup(vocbase, name, returnColPtr);`
in that createSystem method. (Collections.cpp, Row: 437)